### PR TITLE
Run regression test directories in parallel during CI

### PR DESCRIPTION
**Pull request by chrisr-diffblue**
Add some course-grained parallelism when running regression tests during CI. Fine-grained parallelism that is built into tests.pl doesn't appear to be reliable yet, but enabling separate groups of regression tests to run in parallel should give some speed up. Most travis jobs run on dual-core machines, so using that as the number of parallel runs.

### CODE REVIEW

**Review by forejtv — APPROVED** [review]
I'd recommend that @zemanlx also looks at this (unless he's done so already).

**Comment by chrisr-diffblue** [comment]
@forejtv - Sure - I've already had a chat with @zemanlx but will get him to review this PR if I take it further. This is a slightly different approach (a halfway house, really) to the approach he took previously, and I've mainly opened this PR to see what effect it has in CI and to see if its worth taking further.

**Comment by tautschnig** [comment]
I think the `make -C regression test` in line 246 would also benefit from this optimisation (attempt).

**Comment by chrisr-diffblue** [comment]
Indeed, I was just looking at that too. I'll see what effect it has on the cmake builds first and then extend it to the make based builds if it makes enough difference.

**Comment by chrisr-diffblue** [comment]
This seems to save a couple of minutes per-job on the regression tests, so it's probably worth using. Slightly hard to tell because the travis platform isn't terribly predictable (I've seen OSX jobs improve on one run, be slower on the next, etc) - but on average I think it's an improvement.

**Comment by tautschnig** [comment]
I'd say this is just pending a review by @zemanlx?

**Review by peterschrammel — APPROVED** [review]
@zemanlx will be back on Friday.

File: .travis.yml
Comment (zemanlx) [inline .travis.yml]: Have you tried to build it with `-j2` or `-j3` it may be quicker on Travis 2 core machines?

File: .travis.yml
Comment (zemanlx) [inline .travis.yml]: Looks like `ctest` is mixing an output of tests. I also do not know what happens when some test fail? Does it stop or it continues and exits with an error at the end?

**Review by zemanlx — APPROVED** [review]
Every second count so I am happy to approve and tune in next PR.

File: .travis.yml
Comment (chrisr-diffblue) [reply .travis.yml]: I haven't played with the build yet , though I was wondering about the choice of -j4 too, considering the machines we run the builds on. It may be that we'll benefit from different choices on different platforms here, but I'll do some experiments under a different PR.

File: .travis.yml
Comment (chrisr-diffblue) [reply .travis.yml]: Good spot on the interleaving, thanks. My understanding is that ctests will stop and report the error after all running tests complete (i.e. it won't start running new tests after the failure, but in-flight tests will complete). I'm looking into how to deal with the interleaving question now and if possible, will update this PR with a fix for review before merging.

Comment (zemanlx) [reply .travis.yml]: one more thing (Maybe for different PR) cmake osx build .11 do not have ccache

Comment (chrisr-diffblue) [reply .travis.yml]: Ah yes, thats interesting. I'll look at getting ccache going in another PR. It was also pointed out to me recently that the ccache stats for cbmc travis builds don't look as good (i.e. more cache misses) than perhaps we might have expected, so I'll take a look at that in the same PR.

Comment (chrisr-diffblue) [reply .travis.yml]: I haven't found a way to make ctest group it's output more neatly, but it does prefix each output line with a number, so that you can still disambiguate which lines relate to which test directory. While it's not perfect, I think the time saving is probably worth it. I'll be happy to revert this though if anyone finds it too onerous to when debugging any failing tests.

**Comment by chrisr-diffblue** [comment]
I've not heard any howls of complaint, so I'm going to merge this in. If anyone then comes out of the woodwork I'll revert.

**Comment by chrisr-diffblue** [comment]
Sigh, had to rebase, but no other changes. Will merge after CI is green.

**Pull request merged**
